parquet-hadoop/src/main/java/org/apache/parquet/hadoop/InternalParquetRecordWriter.java

```diff
@@ -139,21 +145,41 @@ public long getDataSize() {
 
   private void checkBlockSizeReached() throws IOException {
     if (recordCount >= recordCountForNextMemCheck) { // checking the memory size is relatively expensive, so let's not do it for every record.
+      boolean startNewRowGroup = false;
       long memSize = columnStore.getBufferedSize();
       long recordSize = memSize / recordCount;
-      // flush the row group if it is within ~2 records of the limit
-      // it is much better to be slightly under size than to be over at all
-      if (memSize > (nextRowGroupSize - 2 * recordSize)) {
-        LOG.info("mem size {} > {}: flushing {} records to disk.", memSize, nextRowGroupSize, recordCount);
+      // Write the last pages once we reach the target row group size. This will
+      // encode and compress the pages, which allows us to measure their size
+      // more precisely.
+      if (memSize > nextRowGroupSize) {
+        columnStore.writePages();
+        final long newMemSize = columnStore.getBufferedSize();
+        LOG.debug("mem size {} > {}: closing pages. Mem size after closing pages: {}. Min size for starting new row group: {}",
+          memSize, nextRowGroupSize, newMemSize, minSizeForAlignment);
+        memSize = newMemSize;
+        recordSize = memSize / recordCount;
+        // Start a new group if we are inside the padding area.
+        if (memSize > minSizeForAlignment) {
+          startNewRowGroup = true;
+          LOG.debug("mem size {} > {}: flushing {} records to disk.", memSize, minSizeForAlignment, recordCount);
+        }
+      }
+      if (startNewRowGroup) {
```

Comment:
nit: don't really like this flagging concept in the same method. I would suggest putting the related code to the place where the flag is true.

Reply:
I restructured the code, but now I have a return inside the middle of the function, personally I find this more confusing.

Reply:
I don't have problem with the return. But it is more a personal opinion. I let you choose which one you prefer.